devtools/zenodo/zenodo_data_release.py

```diff
@@ -0,0 +1,295 @@
+#!/usr/bin/env python
+"""Script to sync a directory up to Zenodo."""
+import datetime
+import logging
+import os
+from collections.abc import Iterable
+from dataclasses import dataclass
+from pathlib import Path
+from typing import BinaryIO
+
+import click
+import requests
+
+SANDBOX = "sandbox"
+PRODUCTION = "production"
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger()
+
+
+class ZenodoClient:
+    """Thin wrapper over Zenodo REST API.
+
+    Mostly legacy calls (https://developers.zenodo.org/) (archive:
+    https://web.archive.org/web/20231212025359/https://developers.zenodo.org/)
+    but due to inconsistent behavior of legacy API on sandbox environment, we
+    need some of the unreleased new API endpoints too:
+    https://inveniordm.docs.cern.ch/reference/rest_api_drafts_records/
+    """
+
+    def __init__(self, env: str):
+        """Constructor.
+
+        Args:
+            env: "sandbox" or "production".
+        """
+        if env == SANDBOX:
+            self.base_url = "https://sandbox.zenodo.org/api"
+            self.auth_headers = {
+                "Authorization": f"Bearer {os.environ['ZENODO_SANDBOX_TOKEN_PUBLISH']}"
+            }
+        elif env == PRODUCTION:
+            self.base_url = "https://zenodo.org/api"
+            self.auth_headers = {
+                "Authorization": f"Bearer {os.environ['ZENODO_TOKEN_PUBLISH']}"
+            }
+        else:
+            raise ValueError(
+                f"Got unexpected {env=}, expected {SANDBOX} or {PRODUCTION}"
+            )
+
+    def get_deposition(self, deposition_id: int) -> requests.Response:
+        """LEGACY API: Get JSON describing a deposition.
+
+        Depositions can be published *or* unpublished.
+        """
+        return requests.get(
+            f"{self.base_url}/deposit/depositions/{deposition_id}",
+            headers=self.auth_headers,
+            timeout=10,
+        )
+
+    def get_record(self, record_id: int) -> requests.Response:
+        """NEW API: Get JSON describing a record.
+
+        All records are published records.
+        """
+        return requests.get(
+            f"{self.base_url}/records/{record_id}",
+            headers=self.auth_headers,
+            timeout=10,
+        )
+
+    def new_record_version(self, record_id: int) -> requests.Response:
+        """NEW API: get or create the draft associated with a record ID.
+
+        Finds the latest record in the concept that record_id points to, and
+        makes a new version unless one exists already.
+        """
+        return requests.post(
+            f"{self.base_url}/records/{record_id}/versions",
+            headers=self.auth_headers,
+            timeout=10,
+        )
+
+    def update_deposition_metadata(
+        self, deposition_id: int, metadata: dict
```

Comment:
It would be nice to have a Pydantic model or dataclass that defines what fields we want to require be populated in the Zenodo metadata. It would help make sure we're filling this out "completely" for our purposes.

The only element that I think we may need to change substantively between releases or generate somehow is the description, which I think is HTML, and would be nice to have a Jinja template for that we can fill in based on metadata like the start/end dates of the datasets that are included in the release.

Reply:
Yeah, I sort of wanted to avoid this because I didn't want to make any more promises about the Zenodo API & attendant data shapes than I had to. But I think the promises are being made regardless, and I can codify them with Pydantic models.